File: applicative/simple-applicative-hs/src/SimpleApplicative.hs
module SimpleApplicative
  (
    SimpleApplicative(..)
  ) where

data SimpleApplicative a = SimpleApplicative a
  deriving(Show, Ord, Eq)


instance Functor SimpleApplicative where
  fmap g (SimpleApplicative x) = SimpleApplicative $ g x

instance Applicative SimpleApplicative where
  pure x = SimpleApplicative x
  (SimpleApplicative g) <*> (SimpleApplicative x) = SimpleApplicative (g x)

File: applicative/simple-applicative-ts/src/simple-applicative.ts
export class SimpleApplicative<T> {
    constructor(
        readonly value : T
    ) {}

    fmap<V>( fun: (x: T) => V )
    : SimpleApplicative<V> {
        return fmap(fun, this);
    }
}

export function fromValue<T>(value : T)
: SimpleApplicative<T> {
    return new SimpleApplicative(value);
}

export function fmap<T, V>( fun: (x: T) => V, sa: SimpleApplicative<T> )
: SimpleApplicative<V> {
    return fromValue( fun(sa.value) );
}